In [24]:
!pip install transformers torch torchvision torchaudio tensorflow 


[notice] A new release of pip available: 22.3.1 -> 25.3
[notice] To update, run: C:\Users\Reanna Netto\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip


In [25]:
from transformers import pipeline, set_seed, GPT2Tokenizer

In [26]:
import os
import nltk

In [27]:
file_path = r"C:\Users\Reanna Netto\unit 1.txt"

In [28]:
try:
    with open(file_path, "r", encoding="utf-8") as f:
        text = f.read()
    print("File loaded successfully!")
except FileNotFoundError:
    print(f"Error: '{file_path}' not found.")

File loaded successfully!


# Experiment 1: Text Generation

In [29]:
set_seed(45)

In [30]:
prompt = "The future of AI is"

### BERT

In [31]:
fast_generator = pipeline('text-generation', model="bert-base-uncased")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`
Loading weights: 100%|██████████| 202/202 [00:00<00:00, 227.25it/s, Materializing param=cls.predictions.transform.dense.weight]                 
BertLMHeadModel LOAD REPORT from: bert-base-uncased
Key                         | Status     |  | 
----------------------------+------------+--+-
bert.pooler.dense.weight    | UNEXPECTED |  | 
cls.seq_relationship.weight | UNEXPECTED |  | 
cls.seq_relationship.bias   | UNEXPECTED |  | 
bert.pooler.dense.bias      | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Both `max_new_tokens` (=256) and `max_length`(=25) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


The future of AI is. as housing to " " " " " " ( ) ( " (, - ( ) ( " " ". it out all there way way on being so jefferson long beth for " ( ( " ( ( )ach, " ( ( ". but and and - - - - ( as more actually so " ". of the and and and and ". it more actually, - - ( ) ( ". are and and - ( ( ". it but who the to is'] " ( ) in ". it that as that many so so jefferson is'" ( " ( " ( " ( ( ( ) ) (, that those being " " " ( ( -, that those being that as that it out all shit to ". and and - - - ( " ".. which they with is and " ( ". it this and and " and and ". '!!!!!! j (. an a a the were me to himxt in most em it that a ( ) ) ) ( as and in it it many so the the the do who now will so law them often often " siroch - " ( / the rest with success storm with first a and.,ne general like government general is '


### RoBERTa

In [32]:
fast_generator = pipeline('text-generation', model="roberta-base")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`
Loading weights: 100%|██████████| 202/202 [00:01<00:00, 194.88it/s, Materializing param=roberta.encoder.layer.11.output.dense.weight]              
RobertaForCausalLM LOAD REPORT from: roberta-base
Key                             | Status     |  | 
--------------------------------+------------+--+-
roberta.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Both `max_new_tokens` (=256) and `max_length`(=25) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


The future of AI is


### BART

In [33]:
fast_generator = pipeline('text-generation', model="facebook/bart-base")
output_fast = fast_generator(prompt, max_length=25, num_return_sequences=1)
print(output_fast[0]['generated_text'])

Loading weights: 100%|██████████| 159/159 [00:00<00:00, 239.64it/s, Materializing param=model.decoder.layers.5.self_attn_layer_norm.weight]   
This checkpoint seem corrupted. The tied weights mapping for this model specifies to tie model.decoder.embed_tokens.weight to lm_head.weight, but both are absent from the checkpoint, and we could not find another related tied weight for those keys
BartForCausalLM LOAD REPORT from: facebook/bart-base
Key                                                           | Status     | 
--------------------------------------------------------------+------------+-
encoder.layernorm_embedding.bias                              | UNEXPECTED | 
encoder.layers.{0, 1, 2, 3, 4, 5}.self_attn_layer_norm.weight | UNEXPECTED | 
encoder.layers.{0, 1, 2, 3, 4, 5}.final_layer_norm.bias       | UNEXPECTED | 
encoder.layers.{0, 1, 2, 3, 4, 5}.fc1.weight                  | UNEXPECTED | 
encoder.layers.{0, 1, 2, 3, 4, 5}.fc1.bias                    | UNEXPECTED | 
encoder.la

The future of AI is greetingsuff nanotimeout Laur Airl上上gling defer defer上上 nicerSay nicer Vest Vest_(til Vest prize Vest_( Tenn Di prize defer answersPalest mediocre Boyd defer deferstudentstudentstudent JihadTipoggle orientationstudentTip上 departing Laurcon mediocre defer_(_(_(.) nicerometerometer Jihad�Palest 380 orientation Jihadogglestudentoggle residentsstudentstudentoggleogglestudent orientationmansoggleoggleoggleometeroggle orientation orientation orientationStarting Laur Laur Psychological Laur Laur RIGHT RIGHTиometer.)oggleoggle nicer Psychological Wor deferogglestudent391oggle prizestudentoggle nightmaresoggleoggle prizeoggleoggle CTRometerPalestoggleиoggleoggle Worgling orientation orientationoggleometer.) departingoggleshapeshifteroggleoggle bedsoggleoggle haz.) orientation orientation�oggleoggle Heb orientation haz PsychologicalogglePalest orientation Jihad Heboggleshapeshifter PsychologicalTipUntitledoggleogglethatoggleogglePalestoggleoggle residents391oggle Heb上Palestog

# Experiment 2: Masked Language Modeling (Missing Word)

### BERT

In [34]:
mask_filler = pipeline("fill-mask", model="bert-base-uncased")

Loading weights: 100%|██████████| 202/202 [00:00<00:00, 237.37it/s, Materializing param=cls.predictions.transform.dense.weight]                 
BertForMaskedLM LOAD REPORT from: bert-base-uncased
Key                         | Status     |  | 
----------------------------+------------+--+-
bert.pooler.dense.weight    | UNEXPECTED |  | 
cls.seq_relationship.weight | UNEXPECTED |  | 
cls.seq_relationship.bias   | UNEXPECTED |  | 
bert.pooler.dense.bias      | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [35]:
masked_sentence = "The goal of Generative AI is to [MASK] new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

create: 0.54
generate: 0.16
produce: 0.05
develop: 0.04
add: 0.02


### RoBERTa

In [36]:
mask_filler = pipeline("fill-mask", model="roberta-base")

Loading weights: 100%|██████████| 202/202 [00:00<00:00, 203.87it/s, Materializing param=roberta.encoder.layer.11.output.dense.weight]              
RobertaForMaskedLM LOAD REPORT from: roberta-base
Key                             | Status     |  | 
--------------------------------+------------+--+-
roberta.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [37]:
masked_sentence = "The goal of Generative AI is to <mask> new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

 generate: 0.37
 create: 0.37
 discover: 0.08
 find: 0.02
 provide: 0.02


### BART

In [38]:
mask_filler = pipeline("fill-mask", model="facebook/bart-base")  

Loading weights: 100%|██████████| 259/259 [00:01<00:00, 227.71it/s, Materializing param=model.shared.weight]                                  


In [39]:
masked_sentence = "The goal of Generative AI is to <mask> new content."
preds = mask_filler(masked_sentence)

for p in preds:
    print(f"{p['token_str']}: {p['score']:.2f}")

 create: 0.07
 help: 0.07
 provide: 0.06
 enable: 0.04
 improve: 0.03


# Experiment 3: Question Answering

In [40]:
question = "What are the risks?"
context = "Generative AI poses significant risks such as hallucinations, bias, and deepfakes."

### BERT

In [41]:
qa_pipeline = pipeline("question-answering", model="bert-base-uncased")

Loading weights: 100%|██████████| 197/197 [00:00<00:00, 221.60it/s, Materializing param=bert.encoder.layer.11.output.dense.weight]              
BertForQuestionAnswering LOAD REPORT from: bert-base-uncased
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
bert.pooler.dense.weight                   | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
bert.pooler.dense.bias                     | UNEXPECTED | 
qa_outputs.bias                            | MISSING    | 
qa_outputs.weight                          | MISSING    | 

Notes:
- UNEXPECTED	:can b

In [42]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: , and deepfakes


### RoBERTa

In [43]:
qa_pipeline = pipeline("question-answering", model="roberta-base")

Loading weights: 100%|██████████| 197/197 [00:00<00:00, 210.95it/s, Materializing param=roberta.encoder.layer.11.output.dense.weight]              
RobertaForQuestionAnswering LOAD REPORT from: roberta-base
Key                             | Status     | 
--------------------------------+------------+-
lm_head.dense.weight            | UNEXPECTED | 
lm_head.layer_norm.weight       | UNEXPECTED | 
lm_head.bias                    | UNEXPECTED | 
roberta.embeddings.position_ids | UNEXPECTED | 
lm_head.dense.bias              | UNEXPECTED | 
lm_head.layer_norm.bias         | UNEXPECTED | 
qa_outputs.bias                 | MISSING    | 
qa_outputs.weight               | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


In [44]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: Generative AI poses significant risks such


### BART

In [45]:
qa_pipeline = pipeline("question-answering", model="facebook/bart-base")

Loading weights: 100%|██████████| 259/259 [00:00<00:00, 300.73it/s, Materializing param=model.shared.weight]                                  
BartForQuestionAnswering LOAD REPORT from: facebook/bart-base
Key               | Status  | 
------------------+---------+-
qa_outputs.bias   | MISSING | 
qa_outputs.weight | MISSING | 

Notes:
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


In [46]:
res = qa_pipeline(question=question, context=context)
print(f"Q: {question}")
print(f"A: {res['answer']}")

Q: What are the risks?
A: , and deepfakes


# Observation Table

| Task | Model | Classification (Success/Failure) | Observation (What actually happened?) | Why did this happen? (Architectural Reason) |
| :--- | :--- | :--- | :--- | :--- |
| **Generation** | BERT | *Failure* | *Generated highly incoherent text with repeated punctuation, symbols, and random fragments instead of meaningful sentences.* | *BERT is an encoder-only model trained with Masked Language Modeling and is not designed for next-token (causal) prediction.* |
|  | RoBERTa | *Failure* | *Returned only the input prompt “The future of AI is” without any continuation.* | *RoBERTa is also encoder-only and cannot autoregressively generate new tokens.* |
|  | BART | *Failure* | *Produced extremely long, noisy, and random-looking text containing meaningless words, symbols, and mixed-language fragments.* | *BART is an encoder–decoder model, but the text-generation pipeline expects a decoder-only causal language model, causing architectural misuse.* |
| **Fill-Mask** | BERT | *Success* | *Predicted create (0.54), generate (0.16), produce (0.05).* | *BERT is explicitly trained for Masked Language Modeling (MLM), making it effective at predicting missing tokens.* |
|  | RoBERTa | *Success* | *Predicted generate (0.37), create (0.37), discover (0.08).* | *RoBERTa is trained using MLM with improved optimization and larger pretraining data than BERT.* |
|  | BART | *Partial Success* | *Predicted more generic words like create (0.07), help (0.07), provide (0.06).* | *BART is trained using denoising autoencoding and is not optimized for single-token mask prediction.* |
| **QA** | BERT | *Partial Success* | *Returned incomplete answer fragment: “, and deepfakes”.* | *Base BERT is not fine-tuned for extractive Question Answering, leading to weak span prediction.* |
|  | RoBERTa | *Partial Success* | *Returned truncated answer: “Generative AI poses significant risks such”.* | *Although RoBERTa has strong representations, lack of QA fine-tuning causes incomplete answer spans.* |
|  | BART | *Failure* | *Failed to return a clear or complete answer due to missing QA head parameters.* | *BART-base is primarily designed for sequence-to-sequence generation and requires explicit fine-tuning for extractive QA tasks.* |
